### DISCUSSION TITLE
Performance Logs to Driver()

### DISCUSSION THREAD

**Discussion by Kamran-ov** [seleniumbase/SeleniumBase#2150] [Q&A] [answered]
Hi, 

Can you add an argument for Chrome Driver creation that will add the below capability to Driver?

chrome_options.set_capability('goog:loggingPrefs', {'performance': 'ALL'})

I am trying to do the below 

```
from seleniumbase import Driver
# Create Driver and request Page
driver = Driver(uc_cdp=True, incognito=True, headless=False)
driver.get("https://nowsecure.nl/#relax")
# Get Logs to retrieve AJAX Requests
logs = driver.get_log('performance')
```

**Comment by mdmintz** [comment, marked as answer]
That option is already there for `uc_cdp` mode, which you already appear to be using:

https://github.com/seleniumbase/SeleniumBase/blob/eea244127cf26daac6ec28a3139382378071e0d5/seleniumbase/undetected/__init__.py#L168

**Reply by Kamran-ov** [reply]
Hi,  Thank you for the quick reply.  

for some reason, it says that there are any performance logs. Please see the screenshot below

![image](https://github.com/seleniumbase/SeleniumBase/assets/50695352/c3e68078-2a8f-4ac5-83b5-12606531e19c)

is there any other argument I need to pass or any other way of getting those logs ?

**Reply by mdmintz** [reply]
I think I see the issue. I'm looking into it.

**Reply by mdmintz** [reply]
Try:

```python
from seleniumbase import Driver

driver = Driver(undetectable=True, uc_cdp_events=True)
try:
    driver.get("https://nowsecure.nl/#relax")
    driver.add_cdp_listener(
        "Network.requestWillBeSentExtraInfo",
        lambda data: print(pformat(data))
    )
    driver.refresh()
    driver.sleep(1)
finally:
    driver.quit()
```

(The `uc_cdp` shortcut for `uc_cdp_events` was broken, but it'll be fixed in the next release.)

And apparently `get_log()` can't be used directly. (That method is part of `selenium`, not `seleniumbase`, so I can't do much about that.)

**Reply by Kamran-ov** [reply]
Hi,  Thank you for checking. 
I will wait for the next release then. 

Regarding get_log(). it still works with the driver SleniumBase created. It is just missing proper addition of performance logs. 

when you add 'goog:loggingPrefs' to the options,  you also add browser logs.
For example,  I have tried to get browser logs and they return without any error (below screenshot)

Something is wrong specifically on performance logs. Maybe when you fix ux_cdp_events, this one will also get fixed.

![image](https://github.com/seleniumbase/SeleniumBase/assets/50695352/ec15d073-eb69-4b3d-a6bf-bbff617b6c3d)

**Reply by Kamran-ov** [reply]
@mdmintz,  just for you to know that after the new release,  driver.get_log('performance') works fine.
Thank you again.

**Reply by Kamran-ov** [reply]
Hi @mdmintz, I am having an issue with the performance logs again.

it just returns an empty array instead of a list of logs (I am not sure when it stopped working because it has been a long time since I used SeleniumBase)

![image](https://github.com/seleniumbase/SeleniumBase/assets/50695352/c3064bed-ac73-4e15-87bf-486cd6114edf)

Do you have any suggestions on how I can fix it? 

BTW,  I have the last version of Selenium and SeleniumBase installed.

**Reply by mdmintz** [reply]
Try the script below. Not sure what happened, but this format works. Maybe something to do with thread locking in non-pytest runs.

```python
from pprint import pformat
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__, "--uc", "--uc-cdp", "-s")


class CDPTests(BaseCase):
    def add_cdp_listener(self):
        # (To print everything, use "*". Otherwise select specific headers.)
        # self.driver.add_cdp_listener("*", lambda data: print(pformat(data)))
        self.driver.add_cdp_listener(
            "Network.requestWillBeSentExtraInfo",
            lambda data: print(pformat(data))
        )

    def verify_success(self):
        self.assert_text("OH YEAH, you passed!", "h1", timeout=6.25)
        self.sleep(1)

    def fail_me(self):
        self.fail('Selenium was detected! Try using: "pytest --uc"')

    def test_display_cdp_events(self):
        if not (self.undetectable and self.uc_cdp_events):
            self.get_new_driver(undetectable=True, uc_cdp_events=True)
        self.driver.get("https://nowsecure.nl/#relax")
        try:
            self.verify_success()
        except Exception:
            self.clear_all_cookies()
            self.get_new_driver(undetectable=True, uc_cdp_events=True)
            self.driver.get("https://nowsecure.nl/#relax")
            try:
                self.verify_success()
            except Exception:
                if self.is_element_visible('iframe[src*="challenge"]'):
                    with self.frame_switch('iframe[src*="challenge"]'):
                        self.click("span.mark")
                else:
                    self.fail_me()
                try:
                    self.verify_success()
                except Exception:
                    self.fail_me()
        self.add_cdp_listener()
        self.refresh()
        self.sleep(1)
```

**Reply by Kamran-ov** [reply]
Hi, Should I run it in a specific environment or path?
Running it in Jupiter Notebook.
I am getting the below error :

```
NameError                                 Traceback (most recent call last)
Input In [1], in <cell line: 3>()
      1 from pprint import pformat
      2 from seleniumbase import BaseCase
----> 3 BaseCase.main(__name__, __file__, "--uc", "--uc-cdp", "-s")
      6 class CDPTests(BaseCase):
      7     def add_cdp_listener(self):
      8         # (To print everything, use "*". Otherwise select specific headers.)
      9         # self.driver.add_cdp_listener("*", lambda data: print(pformat(data)))

NameError: name '__file__' is not defined
```

**Reply by mdmintz** [reply]
That has to be run from an actual Python file, using `pytest`.

**Reply by Kamran-ov** [reply]
@mdmintz , The result of the pytest:

```
(base) C:\Users\kamranov\Desktop\test_SB>pytest testsb.py
================================================= test session starts =================================================
platform win32 -- Python 3.9.12, pytest-7.4.2, pluggy-1.3.0
rootdir: C:\Users\kamranov\Desktop\test_SB
plugins: anyio-3.5.0, html-2.0.1, metadata-3.0.0, ordering-0.6, rerunfailures-12.0, xdist-3.3.1, seleniumbase-4.19.1
collected 1 item

testsb.py .                                                                                                      [100%]

================================================= 1 passed in 17.46s ==================================================
```

**Reply by mdmintz** [reply]
Use `pytest -s` to see the output of `print()` statements.

**Reply by Kamran-ov** [reply]
```
================================================= test session starts =================================================
platform win32 -- Python 3.9.12, pytest-7.4.2, pluggy-1.3.0
rootdir: C:\Users\kamranov\Desktop\test_SB
plugins: anyio-3.5.0, html-2.0.1, metadata-3.0.0, ordering-0.6, rerunfailures-12.0, xdist-3.3.1, seleniumbase-4.19.1
collected 1 item

testsb.py {'method': 'Network.requestWillBeSentExtraInfo',
 'params': {'associatedCookies': [{'blockedReasons': [],
                                   'cookie': {'domain': '.nowsecure.nl',
                                              'expires': 1729954310.957342,
                                              'httpOnly': True,
                                              'name': 'cf_clearance',
                                              'path': '/',
                                              'priority': 'Medium',
                                              'sameParty': False,
                                              'sameSite': 'None',
                                              'secure': True,
                                              'session': False,
                                              'size': 114,
                                              'sourcePort': 443,
                                              'sourceScheme': 'Secure',
                                              'value': 'z9h9dHqtTMUuspX9dkBQ3Oq2hDBUMBDNZT9bs_Yda6k-1698418313-0-1-60cccb76.a208d7c9.833fdec3-160.2.1698418313'}}],
            'connectTiming': {'requestTime': 30682.581271},
            'headers': {':authority': 'nowsecure.nl',
                        ':method': 'POST',
                        ':path': '/',
                        ':scheme': 'https',
                        'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
                        'accept-encoding': 'gzip, deflate, br',
                        'accept-language': 'en-US,en;q=0.9',
                        'cache-control': 'max-age=0',
                        'content-length': '3361',
                        'content-type': 'application/x-www-form-urlencoded',
                        'cookie': 'cf_clearance=z9h9dHqtTMUuspX9dkBQ3Oq2hDBUMBDNZT9bs_Yda6k-1698418313-0-1-60cccb76.a208d7c9.833fdec3-160.2.1698418313',
                        'if-modified-since': 'Fri, 27 May 2022 01:42:42 GMT',
                        'origin': 'https://nowsecure.nl',
                        'referer': 'https://nowsecure.nl/?__cf_chl_tk=JNJApPHPFYkka0AhIEPFzrEXdVasNkC0Wg6s1o9TMCA-1698418309-0-gaNycGzNC5A',
                        'sec-ch-ua': '"Chromium";v="118", "Google '
                                     'Chrome";v="118", "Not=A?Brand";v="99"',
                        'sec-ch-ua-mobile': '?0',
                        'sec-ch-ua-platform': '"Windows"',
                        'sec-fetch-dest': 'document',
                        'sec-fetch-mode': 'navigate',
                        'sec-fetch-site': 'same-origin',
                        'sec-fetch-user': '?1',
                        'upgrade-insecure-requests': '1',
                        'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; '
                                      'x64) AppleWebKit/537.36 (KHTML, like '
                                      'Gecko) Chrome/118.0.0.0 Safari/537.36'},
            'requestId': 'E72885C0A768D0696B2AB352ABC41E79',
            'siteHasCookieInOtherPartition': False}}
.

================================================= 1 passed in 16.70s ==================================================
```

**Reply by mdmintz** [reply]
Great. Looks like it’s working. If any more issues, upgrade to the latest version of SeleniumBase.

**Reply by MovingW** [reply]
@mdmintz Hi, I found a weird issue. Code is as below:

<img width="401" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/701c7fac-e9b3-469f-9569-2986f9af2e7b">


If I execute get_log as soon as I finished the get method, I would get the full performance log. But if I add a time.sleep after the get method, I would got nothing. That means, if I couldn't execute get_log in time, all logs would be gone.
I get logs in certain loop, not after every get(), so that I could NOT get logs right after every get().

BTW, I was running the latest seleniumbase.

I was wondering if it's caused by some kind of listener and how to avoid the unintended losing of performance logs. 

Thx.

**Reply by Kamran-ov** [reply]
@MovingW if you are doing sleep after get_log('performance),  then it is normal that you don't get any logs anymore. 
Selenium deletes performance logs immediately after calling the get_log() method. 

So, please try to 
time.sleep(1000) 
webdriver.get_log('performance')

the logs should still be there. 

BTW it is not related to seleniumbase, but how Selenium is built

**Reply by Kamran-ov** [reply]
@mdmintz, yes I just updated the seleniumbase to the latest version. 
I get the performance logs indeed.  thank you again for checking.

**Reply by MovingW** [reply]
Hi @mdmintz @Kamran-ov 

I understood that all logs would be cleared after executing get_log. 
Sorry for the misleading code above. Please refer to the below code. I add a **5 seconds pause** deliberately before executing the webdriver.get_log('performance'), and they all **return nothing.**

I think this issue is from seleniumbase, for this code worked when I used the standard selenium or undetected-chromedriver.

<img width="600" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/06362a3b-7091-495d-a889-b05f3d0d763a">

**### I guess this issue result from the cdp_listener , for the cdp_listener would fetch all logs right after every get request.
How could I turn off the listener but not affect the use of  `webdriver.get_log('performance')`?  I mean, I don't need to get logs after every get, but I need to get all the performance logs together _manually_ at some time.**

**Reply by MovingW** [reply]
Hi @mdmintz 

I understood how this issue occurred. 
I set the cdp_listener parameter to be `uc_cdp=True` only because I need the below setting to execute 'webdriver.get_log('performance')'. But it also bring another issue for me, the regular fetching of performance logs. 

Is there any other way that I could set the desired_capabilities ` "goog:loggingPrefs", {"performance": "ALL", "browser": "ALL"}` but not turn on the regular listeners? 
What I need is I could get all performance logs `webdriver.get_log('performance')` at the right time I need, instead of after every get(url) request.

Thx.

<img width="545" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/af154608-fe87-437a-8600-eb24f22a1287">

**Reply by mdmintz** [reply]
What do you mean by not turning on the regular listeners? That line is here: https://github.com/seleniumbase/SeleniumBase/blob/39cf213bded5f5873c93f76e31586ac4f40948e2/seleniumbase/undetected/__init__.py#L166

**Reply by MovingW** [reply]
Hi @mdmintz ,

Thanks for your quick response.

My need is to execute `webdriver.get_log('performance')` to get all logs. 
<img width="333" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/78dd59c1-971f-49fc-a70c-eaaf8b5cc440">


But at first I met with a issue, which is as below. I guess it's because of the missing setting `bilities  "goog:loggingPrefs", {"performance": "ALL", "browser": "ALL"}`.

<img width="951" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/48db247c-31b2-4639-adc5-595e98a2907c">

So as you mentioned, I turned on the uc_cdp mode. It worked. But if I add a pause before I execute `webdriver.get_log('performance')`, it returned nothing. I guess it's because the logs are fetched by the background cdp_listeners(even if I never add one).
<img width="366" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/332119d6-6bf0-4ed8-b540-8e83e290dbc9">

So my question is, how could I get performance logs only when I execute `webdriver.get_log('performance')`, instead of cdp listeners.

**Reply by mdmintz** [reply]
I'll probably need to make some modification for `self.driver.get_log("performance")` to work when not using `uc_cdp=True`.

**Reply by mdmintz** [reply]
This works:

```python
from seleniumbase import Driver

driver = Driver(uc=True, uc_cdp=True)
try:
    driver.get("https://seleniumbase.io")
    print(driver.get_log("performance"))
finally:
    driver.quit()
```

Prints the logs. It should only work on the currently active window. Otherwise UC Mode would be detected.

**Reply by MovingW** [reply]
@mdmintz 

Wow, glad to hear that! 
I think `driver.get_log("performance")` is a very common usage for the former selenium users. Your upgrade will benefit lots of people.
Thanks for your quick response.

Please let me know if code is updated. Thx.

**Reply by MovingW** [reply]
@mdmintz 

`from seleniumbase import Driver

driver = Driver(uc=True, uc_cdp=True)
try:
    driver.get("https://seleniumbase.io")
    time.sleep(3)
    print(driver.get_log("performance"))
finally:
    driver.quit()`

But if you add a time.sleep(3) before `driver.get_log("performance")`, you'll get nothing.

**Reply by MovingW** [reply]
Hi Michael @mdmintz ,

I've worked out a temporary solution for my issue. I set `uc_cdp=True` and commented out the listen method in Class Reactor. Then it worked. I could execute `driver.get_log("performance")` any time I need. 

Though it's not perfect, it does fix my issue.
Waiting for your perfect upgrade.

<img width="745" alt="image" src="https://github.com/seleniumbase/SeleniumBase/assets/148598755/72d3d937-8cbc-4946-ad1b-d0d38f171944">

**Reply by mdmintz** [reply]
Resolved in `4.20.8` - https://github.com/seleniumbase/SeleniumBase/issues/2220

**Reply by Kamran-ov** [reply]
@mdmintz, thank you very much for this update.  it works perfectly fine with uc=True or uc=None. 
It will help a lot to people working with the logs.

**Reply by Kamran-ov** [reply]
Hi @MovingW and @mdmintz, , 

I just want to put this here for you and for somebody else who is working with logs, and had an issue with the amount of logs returned from the driver.get().

If you use driver.get(url) on the site with antibot systems like Cloudflare etc, 
driver.get_logs('performance')
will return fewer logs than what is actually happening in the network activity tab of the browser.

However, if you use driver.default_get(url),
it will have all the logs without deleting any.

I believe it is because of tab changes etc. when .get() is done with SeleniumBase since it is not doing the default selenium .get() method.

Obviously, you risk getting blocked by using driver.default_get(), since you don't get all the functionalities of Seleniumbase bot bypass features. 
But if you don't have a problem with getting blocked after the initial opening page (in my case, checks are only in the first request of the page), then I would suggest using default_get to get all logs.

**Reply by MovingW** [reply]
Hi @Kamran-ov,

I've never met with similar issues so far. I used seleniumbase to access websites and got logs successfully while I failed using undetected-chromedriver directly.

From my understanding, if you could access the website and all pages were loaded normally, the issues were not caused by the antibot mechanism. If your automation was detected by the antibot system, the server would not respond correctly, so the pages could not be loaded as you expected. And if the pages were loaded as you expected, that meant the server had gave you everything.
